felix/bpf/maps.go

```diff
@@ -299,6 +302,47 @@ func (b *PinnedMap) Delete(k []byte) error {
 	return DeleteMapEntry(b.fd, k, b.ValueSize)
 }
 
+func (b *PinnedMap) copyFromOldMap() error {
+	numEntriesCopied := 0
+	mapMem := make(map[string]struct{})
+	it, err := NewMapIterator(b.oldfd, b.KeySize, b.ValueSize, b.oldSize)
+	if err != nil {
+		return fmt.Errorf("failed to create BPF map iterator: %w", err)
+	}
+	defer func() {
+		err := it.Close()
+		if err != nil {
+			logrus.WithError(err).Panic("Unexpected error from map iterator Close().")
+		}
+	}()
+
+	for {
+		k, v, err := it.Next()
+
+		if err != nil {
+			if err == ErrIterationFinished {
+				return nil
+			}
+			return errors.Errorf("iterating the old map failed: %s", err)
+		}
+
+		if numEntriesCopied == b.MaxEntries {
+			return fmt.Errorf("new map cannot hold all the data from the old map.")
+		}
```

Comment:
What if you just wait for E2BIG error (iirc that is the error returned when the map is full) I wonder if this is always an error. for instance LRU map is likely going to be full but the entries may be expired (say affinity map), perhaps we need the map instance to decide whether the entry needs to be copied. Then you could downsize a full map. Maybe something to note for the next iteration?